File: python/p132.py
# 
# Solution to Project Euler problem 132
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 

import eulerlib, itertools


def compute():
	# Among the integers starting from 2, take the sum of
	# the first 40 integers satisfying the filter condition
	cond = lambda i: eulerlib.is_prime(i) and repunit_mod(10**9, i) == 0
	ans = sum(itertools.islice(filter(cond, itertools.count(2)), 40))
	return str(ans)


def repunit_mod(k, m):
	return (pow(10, k, m * 9) - 1) // 9


if __name__ == "__main__":
	print(compute())

File: haskell/p132.hs
{- 
 - Solution to Project Euler problem 132
 - Copyright (c) Project Nayuki. All rights reserved.
 - 
 - https://www.nayuki.io/page/project-euler-solutions
 - https://github.com/nayuki/Project-Euler-solutions
 -}

import qualified EulerLib


main = putStrLn (show ans)
cond i = EulerLib.isPrime i && (repunitMod (10^9) i) == 0
ans = sum (take 40 (filter cond [2..]))

repunitMod :: Integer -> Integer -> Integer
repunitMod k m = div (powMod 10 k (m * 9) - 1) 9

powMod :: Integer -> Integer -> Integer -> Integer
powMod _ 0 _ = 1
powMod x 1 m = mod x m
powMod x y m = mod ((powMod x (div y 2) m)^2 * (if ((mod y 2) == 0) then 1 else x)) m
